<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.294">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="">

<title>Scripts -</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body>

<div id="quarto-search-results"></div>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Quick Links</h2>
   
  <ul>
  <li><a href="#before-we-begin" id="toc-before-we-begin" class="nav-link active" data-scroll-target="#before-we-begin">0. Before we begin</a>
  <ul class="collapse">
  <li><a href="#overview-dataset-introduction" id="toc-overview-dataset-introduction" class="nav-link" data-scroll-target="#overview-dataset-introduction">Overview &amp; Dataset Introduction</a></li>
  <li><a href="#prerequisites-quick-guide-on-setting-up-rstudio" id="toc-prerequisites-quick-guide-on-setting-up-rstudio" class="nav-link" data-scroll-target="#prerequisites-quick-guide-on-setting-up-rstudio">Prerequisites &amp; Quick Guide on Setting up RStudio</a></li>
  <li><a href="#setting-up-the-environment" id="toc-setting-up-the-environment" class="nav-link" data-scroll-target="#setting-up-the-environment">Setting up the environment</a></li>
  </ul></li>
  <li><a href="#commonly-used-r-packages-in-ampliconmultivariate-analyses" id="toc-commonly-used-r-packages-in-ampliconmultivariate-analyses" class="nav-link" data-scroll-target="#commonly-used-r-packages-in-ampliconmultivariate-analyses">1. Commonly used R packages in amplicon/multivariate analyses</a>
  <ul class="collapse">
  <li><a href="#getting-your-abundance-table-into-r" id="toc-getting-your-abundance-table-into-r" class="nav-link" data-scroll-target="#getting-your-abundance-table-into-r">1.1 Getting your abundance table into R</a></li>
  <li><a href="#packages-for-data-wrangling-data-cleanup" id="toc-packages-for-data-wrangling-data-cleanup" class="nav-link" data-scroll-target="#packages-for-data-wrangling-data-cleanup">1.1 Packages for data wrangling &amp; data cleanup</a></li>
  <li><a href="#packages-for-diversity-statistical-analyses" id="toc-packages-for-diversity-statistical-analyses" class="nav-link" data-scroll-target="#packages-for-diversity-statistical-analyses">1.3 Packages for diversity &amp; statistical analyses</a></li>
  <li><a href="#packages-for-plottingvisualizing-statistical-analyses" id="toc-packages-for-plottingvisualizing-statistical-analyses" class="nav-link" data-scroll-target="#packages-for-plottingvisualizing-statistical-analyses">1.4 Packages for plotting/visualizing statistical analyses</a></li>
  </ul></li>
  <li><a href="#importing-data-used-for-analyses" id="toc-importing-data-used-for-analyses" class="nav-link" data-scroll-target="#importing-data-used-for-analyses">2. Importing data used for analyses</a>
  <ul class="collapse">
  <li><a href="#importing-your-asv-table-into-the-r-environment" id="toc-importing-your-asv-table-into-the-r-environment" class="nav-link" data-scroll-target="#importing-your-asv-table-into-the-r-environment">2.1 Importing your ASV table into the R environment</a></li>
  <li><a href="#importing-taxonomy-and-other-contextual-data" id="toc-importing-taxonomy-and-other-contextual-data" class="nav-link" data-scroll-target="#importing-taxonomy-and-other-contextual-data">2.2 Importing taxonomy and other contextual data</a></li>
  </ul></li>
  <li><a href="#cleaning-up-data" id="toc-cleaning-up-data" class="nav-link" data-scroll-target="#cleaning-up-data">3. Cleaning up data</a></li>
  <li><a href="#data-standardization-normalization-transformation" id="toc-data-standardization-normalization-transformation" class="nav-link" data-scroll-target="#data-standardization-normalization-transformation">4. Data Standardization, Normalization &amp; Transformation</a></li>
  <li><a href="#alpha-diversity" id="toc-alpha-diversity" class="nav-link" data-scroll-target="#alpha-diversity">5. Alpha diversity</a></li>
  <li><a href="#distance-metrices" id="toc-distance-metrices" class="nav-link" data-scroll-target="#distance-metrices">6. Distance Metrices</a></li>
  <li><a href="#ordination-methods" id="toc-ordination-methods" class="nav-link" data-scroll-target="#ordination-methods">7. Ordination Methods</a></li>
  <li><a href="#multivariate-comparisons" id="toc-multivariate-comparisons" class="nav-link" data-scroll-target="#multivariate-comparisons">8. Multivariate Comparisons</a></li>
  <li><a href="#exploring-phylogenetic-diversity" id="toc-exploring-phylogenetic-diversity" class="nav-link" data-scroll-target="#exploring-phylogenetic-diversity">9. Exploring Phylogenetic Diversity</a></li>
  <li><a href="#more-interesting-materials-and-references" id="toc-more-interesting-materials-and-references" class="nav-link" data-scroll-target="#more-interesting-materials-and-references">More interesting materials and references</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">



<section id="nioz-bioinformatics-workshop-2023" class="level2 unnumbered unlisted">
<h2 class="unnumbered unlisted anchored" data-anchor-id="nioz-bioinformatics-workshop-2023">NIOZ Bioinformatics Workshop 2023</h2>
<section id="s8-amplicon-data-analyses-in-r" class="level3 unnumbered unlisted">
<h3 class="unnumbered unlisted anchored" data-anchor-id="s8-amplicon-data-analyses-in-r">S8: Amplicon Data Analyses in R</h3>
<section id="swan-ls-sow-gonçalo-piedade-harry-witte" class="level4 unnumbered unlisted">
<h4 class="unnumbered unlisted anchored" data-anchor-id="swan-ls-sow-gonçalo-piedade-harry-witte">Swan LS Sow | Gonçalo Piedade | Harry Witte</h4>
<section id="march-30-2023" class="level5 unnumbered unlisted">
<h5 class="unnumbered unlisted anchored" data-anchor-id="march-30-2023">March 30, 2023</h5>
</section>
</section>
</section>
</section>
<section id="before-we-begin" class="level1">
<h1>0. Before we begin</h1>
<section id="overview-dataset-introduction" class="level3">
<h3 class="anchored" data-anchor-id="overview-dataset-introduction">Overview &amp; Dataset Introduction</h3>
<p>In this practical, you will learn how to perform basic statistical and multivariate analyses on a dataset generated from 16S amplicon sequences. All analyses will be based on OTU/ASV (.biom) tables (and their corresponding contextual/taxonomy data), that is the output generated from CASCABEL analyses (which you have learned in S5 previously) on the raw sequence data.</p>
<p>The dataset we are using today was generated from samples in this nice paper by <a href="https://www.frontiersin.org/articles/10.3389/fmicb.2021.673553/full">Vaksmaa et.al. (2021)</a></p>
<p><img src="https://user-images.githubusercontent.com/25445935/228508123-37a4d3e6-3d87-4fc9-bd3c-4f8b1d40f1f5.png" class="quarto-discovered-preview-image" height="300"></p>
<p>All analyses will be done in RStudio, either in <em>ada</em> or on your own computer.</p>
<p>The general workflow of amplicon data analyses looks something like this:</p>
<p><img src="https://user-images.githubusercontent.com/25445935/228508141-aa824f33-28ee-4e7f-9cf9-e7c55f5636a2.png" class="img-fluid" width="720">.</p>
<p>You will learn:</p>
<ul>
<li>how to import .biom files into R</li>
<li>dataframe/matrix wrangling to filter and remove unwanted/outlier data points</li>
<li>statistically analyze the “clean” dataset with various multivariate and phylogenetic analyses methods.</li>
</ul>
<p>Please bear in mind that these are just examples to illustrate how the analyses is done. Some of the analyses method may not be the best fit for the example dataset, for various reasons that we will not explore today as it is beyond the scope of this session. Do feel free to swap this up for your own data, and adapt the normalizations, transformations, resemblance measures, ordination methods etc. to suit.</p>
<p>The workflow described above is by no means exhaustive - there is of course more aspects of the data that can be analyzed and methods that can be used, but we will be using that workflow as a guide for the purposes of this tutorial.</p>
</section>
<section id="prerequisites-quick-guide-on-setting-up-rstudio" class="level3">
<h3 class="anchored" data-anchor-id="prerequisites-quick-guide-on-setting-up-rstudio">Prerequisites &amp; Quick Guide on Setting up RStudio</h3>
<ol type="1">
<li><p>If required, connect to NIOZ’s VPN with the Forticlient Console</p></li>
<li><p>On a web browser, key in the following link: <a href="http://ada.nioz.nl/8787" class="uri">http://ada.nioz.nl/8787</a></p></li>
<li><p>Login with your NIOZ username (not email address) and server password (i.e.&nbsp;the same one you use to connect to <em>ada</em>)</p></li>
<li><p>The files required for this analyses are all located in: /export/lv3/scratch/bioinformatics_workshop/S8_9_AmpliconAnalysesInR</p></li>
<li><p>Please copy any script(s) and softlink the required data files to your own working directory for this workshop. You can create your working directory in:</p></li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="sc">/</span>export<span class="sc">/</span>lv3<span class="sc">/</span>scratch<span class="sc">/</span>bioinformatics_workshop<span class="sc">/</span>Users<span class="sc">/</span><span class="er">&lt;</span>yourusernamehere<span class="sc">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="setting-up-the-environment" class="level3">
<h3 class="anchored" data-anchor-id="setting-up-the-environment">Setting up the environment</h3>
<p><i>devtools</i> is a R package that allows you to download R Packages that are in development</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><i>devtools</i> can be hard to install on linux and mac. Usually it requires some tools to be installed via terminal. For more info, see: <a href="https://stackoverflow.com/questions/48469005/install-devtools-package-in-ubuntu-vm">Installing Devtools in Ubuntu</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"devtools"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="commonly-used-r-packages-in-ampliconmultivariate-analyses" class="level1">
<h1>1. Commonly used R packages in amplicon/multivariate analyses</h1>
<p>The following packages are commonly used for community and multivariate analyses in R:</p>
<section id="getting-your-abundance-table-into-r" class="level3">
<h3 class="anchored" data-anchor-id="getting-your-abundance-table-into-r">1.1 Getting your abundance table into R</h3>
<p><i>biom</i> is installed via github with devtools, and allows you to import .biom tables in R</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"biom"</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(biom) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><i>biom</i> might not be compatible with newer versions of R. If you get an error that package is not available, install <i>biomformat</i> instead (also via devtools)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># BiocManager::install("biomformat")</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(biomformat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="packages-for-data-wrangling-data-cleanup" class="level3">
<h3 class="anchored" data-anchor-id="packages-for-data-wrangling-data-cleanup">1.1 Packages for data wrangling &amp; data cleanup</h3>
<p>This was comprehensively covered by Nina Dombrowski in the previous session. If you need a refresher please see the <a href="https://ndombrowski.github.io/Tidyverse_tutorial/#some-background-information"><i>tidyverse</i> tutorial</a> she created.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="packages-for-diversity-statistical-analyses" class="level3">
<h3 class="anchored" data-anchor-id="packages-for-diversity-statistical-analyses">1.3 Packages for diversity &amp; statistical analyses</h3>
<p>The <i>vegan</i> community ecology package is one of the oldest and most widely used package for descriptive community ecology. It has tools and functions for basic diversity analysis, ordination methods and dissimilarity analysis.</p>
<p><i>vegan</i> should already be installed in <i>ada</i>, but if you would like to install it locally, I encourage you to check out the installation options in <a href="https://github.com/vegandevs/vegan">vegan’s Github page</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vegan)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Other packages we will use/recommend are:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Package with a set of tools to facilitate the import, storage, analysis, and graphical display of microbiome census data</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phyloseq)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Package with essential tools for the compositional approach to data analyses </span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(coda.base)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Package with functions for the consistent analysis of compositional data proposed by J. Aitchison et.al.</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(compositions)</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Package for exploratory and Euclidean Methods in environmental science</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ade4)</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Package for exploratory multivariate data analyses</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(FactoMineR)</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Package for ordination and multivariate analyses</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(labdsv)</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Package for spatial analysis, inference on diversity indices</span></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pgirmess)</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Package with (too) many (to list all here) useful functions for data analysis, utility ops., computing sample size/power...</span></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Hmisc) <span class="co"># Harrell Miscellaneous</span></span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Package that treats zeros and non-detects in compositional data sets</span></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(zCompositions)</span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Package for compuing pairwise PERMANOVA</span></span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ecole)</span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Package for differential (relative) abundance analysis of high throughput sequencing count-compositional data</span></span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ALDEx2)</span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a><span class="co"># Package for calculating taxonomic, functional, and phylogenetic diversity of ecological communities as Hill numbers</span></span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(hillR)</span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Pipe-Friendly Framework for Basic Statistical Tests</span></span>
<span id="cb8-38"><a href="#cb8-38" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rstatix)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="packages-for-plottingvisualizing-statistical-analyses" class="level3">
<h3 class="anchored" data-anchor-id="packages-for-plottingvisualizing-statistical-analyses">1.4 Packages for plotting/visualizing statistical analyses</h3>
<p>We will use ggplot (of course!). This was also covered by Nina in an excellent <a href="https://ndombrowski.github.io/Ggplot_tutorial/#playground"><i>ggplot</i> tutorial</a> in the previous session.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Other handy packages are:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Package to automatically Position Non-Overlapping Text Labels with 'ggplot2'</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot2 based publication ready plots</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggpubr)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Package with easy-to-use functions to extract and visualize the output of multivariate data analyses e.g. PCA, CA, MCA,FAMD, MFA, HMFA</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(factoextra)</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Package to visualize correlation matrices</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(corrplot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can checkout more detailed information and vignettes of a specific package:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">browseVignettes</span>(<span class="st">"&lt;nameofthepackage&gt;"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="importing-data-used-for-analyses" class="level1">
<h1>2. Importing data used for analyses</h1>
<p>To view our OTU/ASV table, we will load the .biom file, which is the output file of CASCABEL.</p>
<p><a href="https://biom-format.org/index.html">BIOM</a> stands for <b>B</b>iological <b>O</b>bservation <b>M</b>atrix and is a widely used file format for observation/abundance tables of biological samples in comparative -omics. The file stores both the observation data and corresponding metadata of the observation (e.g.&nbsp;taxonomic classification, metadata of a genome/gene family, pathway etc.)</p>
<p>Usually we use the the .biom file with singletons already removed:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>asvtab.biom <span class="ot">&lt;-</span> <span class="fu">read_biom</span>(<span class="st">"/Users/slssow/Documents/NIOZ/2023_NIOZ/Bioinfo_Workshop_2023/S8_AmpliconDataAnalysisRPt1/NIOZ140_new_result/asvTable_noSingletons.biom"</span>) </span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="co"># change the file path to where your copy of the .biom file is actually located</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="importing-your-asv-table-into-the-r-environment" class="level3">
<h3 class="anchored" data-anchor-id="importing-your-asv-table-into-the-r-environment">2.1 Importing your ASV table into the R environment</h3>
<p>We’ll then proceed to extract the species-site matrix from the .biom file we have loaded into the R environment, and convert it into a format (either a matrix/dataframe) that is easily manipulated in R.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>asvtab.mat<span class="ot">&lt;-</span><span class="fu">as.matrix</span>(<span class="fu">biom_data</span>(asvtab.biom)) <span class="co"># OR</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co"># asvtab.df&lt;-as.data.frame(biom_data(asvtab.biom))</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The first thing we need to do is to check if our ASV table was imported into R properly, with all the samples and ASVs we would expect from our dataset. It isn’t possible to import the ASV table directly as a dataframe, you will get the following error message:</p>
<p><img src="https://user-images.githubusercontent.com/25445935/228508118-2823be17-8312-43d5-bdbf-2ba84ac95001.png" class="img-fluid"></p>
<p>You could do this instead, then you could directly view the number of rows/columns (No.&nbsp;of ASVs/No.&nbsp;of Samples) in the <i>Environment</i> tab of your R studio console:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>asvtab.df<span class="ot">&lt;-</span><span class="fu">as.data.frame</span>(<span class="fu">as.matrix</span>(<span class="fu">biom_data</span>(asvtab.biom)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Or, you could just load your data as a matrix and use one of the following commands to check number of rows/columns and explore your dataset:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(asvtab.mat) <span class="co"># this gives you the number of rows/columns of your matrix</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(asvtab.mat) <span class="co"># this gives you a complex display of the structure of your R object</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(asvtab.mat) <span class="co"># this allows you to view the first 6 rows of your matrix</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(asvtab.mat) <span class="co"># this allows you to view the last 6 rows of your matrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Questions:</p>
<p>1. How many samples are in this dataset?</p>
<details>
<summary>
Show answers
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(asvtab.mat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> num [1:51400, 1:41] 0 8 0 0 6 0 0 11 0 11 ...
 - attr(*, "dimnames")=List of 2
  ..$ : chr [1:51400] "asv.1" "asv.2" "asv.3" "asv.4" ...
  ..$ : chr [1:41] "NIOZ140.1.1" "NIOZ140.1.10" "NIOZ140.1.11" "NIOZ140.1.2" ...</code></pre>
</div>
</div>
<p>There are 41 samples</p>
</details>
<p>2. How many ASVs are in this dataset?</p>
<details>
<summary>
Show answers
</summary>
<p>Same code as above, there are 51400 ASVs.</p>
</details>
<p>3. What is the name of the first and last ASV in this dataset?</p>
<details>
<summary>
Show answers
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>asvtab.mat [<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      NIOZ140.1.1 NIOZ140.1.10 NIOZ140.1.11
asv.1           0            0            0
asv.2           8           57            0
asv.3           0            3            3</code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>asvtab.mat [<span class="dv">51399</span><span class="sc">:</span><span class="dv">51400</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          NIOZ140.1.1 NIOZ140.1.10 NIOZ140.1.11
asv.51399           0            0            0
asv.51400           0            0            0</code></pre>
</div>
</div>
</details>
</section>
<section id="importing-taxonomy-and-other-contextual-data" class="level3">
<h3 class="anchored" data-anchor-id="importing-taxonomy-and-other-contextual-data">2.2 Importing taxonomy and other contextual data</h3>
<p>As mentioned above, the biom file format is also capable of storing taxonomic classification for the ASVs or OTUs. We’ll go ahead and extract the taxonomy for the ASVs as a R matrix:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>taxo.mat <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">observation_metadata</span>(asvtab.biom))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can use the same commands as you did for asvtab.mat to find out the number of ASVs and what the column names are in the taxonomy table.</p>
<p>Questions:</p>
<p>1. How many ASVs are in the taxonomy table?</p>
<details>
<summary>
Show answers
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(taxo.mat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> chr [1:51400, 1:7] "Bacteria" "Bacteria" "Bacteria" "Bacteria" "Bacteria" ...
 - attr(*, "dimnames")=List of 2
  ..$ : chr [1:51400] "asv.1" "asv.2" "asv.3" "asv.4" ...
  ..$ : chr [1:7] "taxonomy1" "taxonomy2" "taxonomy3" "taxonomy4" ...</code></pre>
</div>
</div>
</details>
<p>2. How many taxonomic levels are there in the taxonomy table?</p>
<details>
<summary>
Show answers
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(taxo.mat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      taxonomy1  taxonomy2        taxonomy3             taxonomy4         
asv.1 "Bacteria" "Bacteroidota"   "Bacteroidia"         "Flavobacteriales"
asv.2 "Bacteria" "Bacteroidota"   "Bacteroidia"         "Flavobacteriales"
asv.3 "Bacteria" "Proteobacteria" "Gammaproteobacteria" "Pseudomonadales" 
asv.4 "Bacteria" "Bacteroidota"   "Bacteroidia"         "Flavobacteriales"
asv.5 "Bacteria" "Cyanobacteria"  "Cyanobacteriia"      "Chloroplast"     
asv.6 "Bacteria" "Bacteroidota"   "Bacteroidia"         "Flavobacteriales"
      taxonomy5              taxonomy6    taxonomy7
asv.1 "Flavobacteriaceae"    "Aquimarina" "NA"     
asv.2 "Flavobacteriaceae"    "Gramella"   "NA"     
asv.3 "Saccharospirillaceae" "Oleibacter" "NA"     
asv.4 "Flavobacteriaceae"    "Aquimarina" "NA"     
asv.5 "NA"                   "NA"         "NA"     
asv.6 "Flavobacteriaceae"    "Aquimarina" "NA"     </code></pre>
</div>
</div>
</details>
<p>Notice that the columns names “taxonomy1,2,3…” are not very informative taxonomic ranks. let’s change them with a vector of 7 classical taxonomic ranks (as per the common convention in Silva):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(taxo.mat)<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"Kingdom"</span>, <span class="st">"Phylum"</span>, <span class="st">"Class"</span>,<span class="st">"Order"</span>, <span class="st">"Family"</span>, <span class="st">"Genus"</span>, <span class="st">"Species"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<details>
<summary>
Show the rest of the code:
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(taxo.mat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Kingdom    Phylum           Class                 Order             
asv.1 "Bacteria" "Bacteroidota"   "Bacteroidia"         "Flavobacteriales"
asv.2 "Bacteria" "Bacteroidota"   "Bacteroidia"         "Flavobacteriales"
asv.3 "Bacteria" "Proteobacteria" "Gammaproteobacteria" "Pseudomonadales" 
asv.4 "Bacteria" "Bacteroidota"   "Bacteroidia"         "Flavobacteriales"
asv.5 "Bacteria" "Cyanobacteria"  "Cyanobacteriia"      "Chloroplast"     
asv.6 "Bacteria" "Bacteroidota"   "Bacteroidia"         "Flavobacteriales"
      Family                 Genus        Species
asv.1 "Flavobacteriaceae"    "Aquimarina" "NA"   
asv.2 "Flavobacteriaceae"    "Gramella"   "NA"   
asv.3 "Saccharospirillaceae" "Oleibacter" "NA"   
asv.4 "Flavobacteriaceae"    "Aquimarina" "NA"   
asv.5 "NA"                   "NA"         "NA"   
asv.6 "Flavobacteriaceae"    "Aquimarina" "NA"   </code></pre>
</div>
</div>
</details>
<p>Lastly, we’ll also read in a table with other metadata associated with this dataset. We want more information about the plastics that the microbes are associated with.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>metad.N140 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"/Users/slssow/Documents/NIOZ/2023_NIOZ/Bioinfo_Workshop_2023/S8_AmpliconDataAnalysisRPt1/NIOZ140_new_result/Metadata_N140.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Questions:</p>
<p>1. What types of metadata can we find from the metadata table?</p>
<details>
<summary>
Show answer:
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(metad.N140)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Sample_ID Sample Plastic
1  NIOZ140.1.1  N.1.1      PE
2 NIOZ140.1.10 N.1.10      PP
3 NIOZ140.1.11 N.1.11      PE
4  NIOZ140.1.2  N.1.2      PE
5  NIOZ140.1.3  N.1.3      PE
6  NIOZ140.1.4  N.1.4      PE</code></pre>
</div>
</div>
</details>
<p>2. How many types of plastics are there?</p>
<details>
<summary>
Show answer:
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>metad.N140 <span class="sc">|&gt;</span> <span class="fu">distinct</span>(Plastic) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Plastic
1      PE
2      PP
3      PS
4       X
5      NC</code></pre>
</div>
</div>
</details>
</section>
</section>
<section id="cleaning-up-data" class="level1">
<h1>3. Cleaning up data</h1>
<p>Before doing any statistical or multivariate analyses on our data, we need to clean up our data first to remove outliers, noise and artefacts</p>
<p>If you haven’t already done it above, load the tidyverse/dplyr library as this is handy for data manipulation (as we practiced from the previous session)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>First, let’s remove ASVs with less than 10 reads in total (across all samples)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>asvtab.mat.md10 <span class="ot">&lt;-</span> asvtab.mat[<span class="fu">rowSums</span>(asvtab.mat) <span class="sc">&gt;=</span> <span class="dv">10</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Question: How many ASVs are left?</p>
<details>
<summary>
Show answer:
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(asvtab.mat.md10)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 39296    41</code></pre>
</div>
</div>
</details>
<p>We also want to remove ASVs that are only present in one sample. To do this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>asvtab.mat.md10.pa <span class="ot">&lt;-</span> <span class="fu">decostand</span>(asvtab.mat.md10, <span class="at">method =</span> <span class="st">"pa"</span>, <span class="at">MARGIN =</span> <span class="dv">2</span>) <span class="co"># gives the presence (1) absence (0) of each OTUs/ASVs across samples</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>asvtab.mat.md10.pa <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(asvtab.mat.md10.pa) <span class="co"># convert matrix to a dataframe (df) so I can use rowSums function</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>asvtab.mat.md10.pa<span class="sc">$</span>sumrow <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(asvtab.mat.md10.pa, <span class="at">na.rm =</span> <span class="cn">TRUE</span>) <span class="co"># sum up total samples where ASV was present in a new column called "sumrow"</span></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Now I want to merge column "sumrow" to df asvtab.mat.md10, so asvtab.mat.md10 also needs to be a df</span></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>asvtab.mat.md10 <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(asvtab.mat.md10) </span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge sumrow to df asvtab.mat.md10 with cbind</span></span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>asvtab.mat.md10 <span class="ot">&lt;-</span> <span class="fu">cbind</span>(asvtab.mat.md10, asvtab.mat.md10.pa<span class="sc">$</span>sumrow)</span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Then Use tidyverse (dplyr) to filter out sumrow less than or equal to one</span></span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a>asvtab.mat.md10.md1s <span class="ot">&lt;-</span> asvtab.mat.md10 <span class="sc">|&gt;</span> <span class="fu">filter</span>(asvtab.mat.md10.pa<span class="sc">$</span>sumrow <span class="sc">&gt;</span> <span class="dv">1</span>)</span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-16"><a href="#cb39-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove "asvtab.mat.md10.pa$sumrow" column (we don't need this for further analyses steps)</span></span>
<span id="cb39-17"><a href="#cb39-17" aria-hidden="true" tabindex="-1"></a>asvtab.mat.md10.md1s <span class="ot">&lt;-</span> asvtab.mat.md10.md1s[ <span class="sc">-</span><span class="fu">c</span>(<span class="dv">42</span>) ]</span>
<span id="cb39-18"><a href="#cb39-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-19"><a href="#cb39-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Check dimensions of df to make sure merge went well:</span></span>
<span id="cb39-20"><a href="#cb39-20" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(asvtab.mat.md10.md1s)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 17239    41</code></pre>
</div>
</div>
<p>Question: How many ASVs are left now?</p>
<p>Next, we also want to:</p>
<ul>
<li>Remove unassigned ASVs</li>
<li>Remove ASVs with only * in taxonomy, even at Domain level</li>
<li>Remove ASVs classified as Eukaryote</li>
<li>Remove ASVs classified as Chloroplast and Mitochondria</li>
</ul>
<p>Can you do this on your own?</p>
<details>
<summary>
I’m not ready yet, show me the code:
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># First merge ASV and taxonomy table before proceeding:</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>taxo.mat <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(taxo.mat) <span class="co"># convert taxo table from matrix to df</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>asvtab.mat.md10.md1s.taxo <span class="ot">&lt;-</span> <span class="fu">merge</span> (asvtab.mat.md10.md1s, taxo.mat, <span class="at">by.x =</span> <span class="dv">0</span>, <span class="at">by.y =</span> <span class="dv">0</span>) <span class="co"># merge is inner by default</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Start with 17239 after ASV table cleanup earlier, 49 variables after addition of 8 taxonomy columns</span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove ASVs where Kingdom is not Bacteria</span></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>asvtab.mat.md10.md1s.taxo.clean <span class="ot">&lt;-</span> asvtab.mat.md10.md1s.taxo <span class="sc">|&gt;</span> <span class="fu">filter</span>(Kingdom <span class="sc">==</span> <span class="st">"Bacteria"</span>)</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a><span class="co"># 17094 ASVs remain</span></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove ASVs where Phylum is unassigned (i.e. 'NA')</span></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>asvtab.mat.md10.md1s.taxo.clean <span class="ot">&lt;-</span> asvtab.mat.md10.md1s.taxo.clean <span class="sc">|&gt;</span> <span class="fu">filter</span>(Phylum <span class="sc">!=</span> <span class="st">"NA"</span>)</span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a><span class="co"># 17069 ASVs remain</span></span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove 'Chloroplast' from Order</span></span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a>asvtab.mat.md10.md1s.taxo.clean <span class="ot">&lt;-</span> asvtab.mat.md10.md1s.taxo.clean <span class="sc">|&gt;</span> <span class="fu">filter</span>(Order <span class="sc">!=</span> <span class="st">"Chloroplast"</span>)</span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a><span class="co"># 16401 ASVs remain</span></span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove 'Mitochondria' from Family</span></span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a>asvtab.mat.md10.md1s.taxo.clean <span class="ot">&lt;-</span> asvtab.mat.md10.md1s.taxo.clean <span class="sc">|&gt;</span> <span class="fu">filter</span>(Family <span class="sc">!=</span> <span class="st">"Mitochondria"</span>)</span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a><span class="co"># 15100 ASVs remain</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</details>
<p>Pop quiz: How many ASVs were not Bacteria? And which samples are classifed as plastic X?</p>
<details>
<summary>
Show answer:
</summary>
<p>Number of ASVs that were not Bacteria:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="dv">17239-17094</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 145</code></pre>
</div>
</div>
<p>Samples that are classified as plastic X:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>metad.N140 <span class="sc">|&gt;</span> <span class="fu">filter</span>(Plastic <span class="sc">==</span> <span class="st">"X"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    Sample_ID Sample Plastic
1 NIOZ140.4.4  N.4.4       X
2 NIOZ140.4.5  N.4.5       X</code></pre>
</div>
</div>
</details>
<p>One more thing, we would also want to:</p>
<ul>
<li>Remove sample ‘NIOZ140.NC’, as this is the control</li>
<li>Remove samples classified as unknown plastic X</li>
</ul>
<details>
<summary>
I’m still not ready yet, show me the code:
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Separate clean ASV table from taxonomy, also remove control NIOZ140.NC</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>asvtab.c <span class="ot">&lt;-</span> asvtab.mat.md10.md1s.taxo.clean <span class="sc">|&gt;</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">!</span><span class="fu">c</span>(NIOZ140.NC, Kingdom, Phylum, Class, Order, Family, Genus, Species))</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a><span class="co"># From the pop quiz above we know that the 2 samples classified as plastic X are NIOZ140.4.4 and NIOZ140.4.5</span></span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>asvtab.c.noX <span class="ot">&lt;-</span> asvtab.c <span class="sc">|&gt;</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">!</span><span class="fu">c</span>(NIOZ140.<span class="fl">4.4</span>, NIOZ140.<span class="fl">4.5</span>))</span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Row.names (i.e. ASVId) got shifted to first column, assign them back as index column</span></span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>asvtab.c <span class="ot">&lt;-</span> asvtab.c <span class="sc">|&gt;</span> <span class="fu">column_to_rownames</span>(<span class="at">var =</span> <span class="st">'Row.names'</span>)</span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a><span class="co"># 40 samples, 15100 ASVs</span></span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a>asvtab.c.noX <span class="ot">&lt;-</span> asvtab.c.noX <span class="sc">|&gt;</span> <span class="fu">column_to_rownames</span>(<span class="at">var =</span> <span class="st">'Row.names'</span>)</span>
<span id="cb46-11"><a href="#cb46-11" aria-hidden="true" tabindex="-1"></a><span class="co"># 38 samples, 15100 ASVs</span></span>
<span id="cb46-12"><a href="#cb46-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-13"><a href="#cb46-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Also create a separate clean taxonomy table for later on</span></span>
<span id="cb46-14"><a href="#cb46-14" aria-hidden="true" tabindex="-1"></a>taxotab.c <span class="ot">&lt;-</span> asvtab.mat.md10.md1s.taxo.clean <span class="sc">|&gt;</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">c</span>(Row.names, Kingdom, Phylum, Class, Order, Family, Genus, Species))</span>
<span id="cb46-15"><a href="#cb46-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Row.names (i.e. ASVId) got shifted to first column, assign them back as index column</span></span>
<span id="cb46-16"><a href="#cb46-16" aria-hidden="true" tabindex="-1"></a>taxotab.c <span class="ot">&lt;-</span> taxotab.c <span class="sc">|&gt;</span> <span class="fu">column_to_rownames</span>(<span class="at">var =</span> <span class="st">'Row.names'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</details>
<p>The last thing you want to do in the clean-up step is to check that your datasets (ASV IDs, sample name/IDs and ASV IDs in taxonomy table) match one another. This will facilitate easier comparison and the downstream analyses steps.</p>
<p>To do this, first save the co-occuring column and row names of all the dataframes you will be working with as separate R objects</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(asvtab.c) <span class="co"># colnames = SampleID</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(asvtab.c) <span class="co"># rownames = ASVId</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(taxotab.c) <span class="co"># rownames = ASVId</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(metad.N140) <span class="co"># [1] "Sample_ID" "Sample" "Plastic"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To check that samples IDs in OTU table matches metadata table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(asvtab.c) <span class="sc">%in%</span> metad.N140<span class="sc">$</span>Sample_ID</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
[16] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
[31] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</code></pre>
</div>
</div>
<p>From the comparison above, we know that all the sample IDs in the ASV table can be found in the metadata table, but are they in the same sequence?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># %in% only compares if the elements in the list are the same. Can be used with both lists and vectors</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="co"># == compares if the elements AND the sequence in the list are the same. Can only be used with vectors</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(asvtab.c) <span class="sc">==</span> metad.N140<span class="sc">$</span>Sample_ID</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in colnames(asvtab.c) == metad.N140$Sample_ID: longer object length is
not a multiple of shorter object length</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
[13]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
[25]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
[37]  TRUE  TRUE  TRUE  TRUE FALSE</code></pre>
</div>
</div>
<p>No match! SampleID sequence is right, but we forgot to remove control NIOZ140.NC from the contextual data table.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove control NIOZ140.NC from the contextual data table</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>metad.N140.c <span class="ot">&lt;-</span> metad.N140 <span class="sc">|&gt;</span> <span class="fu">filter</span>(Plastic <span class="sc">!=</span> <span class="st">"NC"</span>)</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Check again that Sample_ID in ASV table matches metadata table</span></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(asvtab.c) <span class="sc">==</span> metad.N140.c<span class="sc">$</span>Sample_ID</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
[16] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
[31] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</code></pre>
</div>
</div>
<p>Also remove plastic X from the contextual data table</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>metad.N140.c.noX <span class="ot">&lt;-</span> metad.N140.c <span class="sc">|&gt;</span> <span class="fu">filter</span>(Plastic <span class="sc">!=</span> <span class="st">"X"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally, let’s check that ASV Ids in ASV table matches the ones in the taxo table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(asvtab.c) <span class="sc">==</span> <span class="fu">rownames</span>(taxotab.c)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Toooooo many ASVs? Try:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">rownames</span>(asvtab.c) <span class="sc">==</span> <span class="fu">rownames</span>(taxotab.c))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Mode    TRUE 
logical   15100 </code></pre>
</div>
</div>
</section>
<section id="data-standardization-normalization-transformation" class="level1">
<h1>4. Data Standardization, Normalization &amp; Transformation</h1>
<p>Hot tips:</p>
<ul>
<li>transformations != normalizations</li>
<li>normalizations recast the data in absolute terms, transformations do not.</li>
<li>transformations downweight the contributions of quantitatively dominant ASVs to the similarities/correlations calculated between samples</li>
<li>The results of a transformation-based analysis must be interpreted with respect to the chosen reference.</li>
</ul>
<section id="species-x-site-a.k.a.-asv-table-normalizations" class="level4">
<h4 class="anchored" data-anchor-id="species-x-site-a.k.a.-asv-table-normalizations">Species x Site (a.k.a. ASV table) normalizations</h4>
<p>The decostand function in vegan can do many of the common standardization/normalization/transformation. To explore all the standardization/transformation methods available:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>?<span class="fu">decostand</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The first few methods described are more ‘old school’, but we will also explore methods deemed more suitable for compositional data analyses.</p>
<p>Methods suitable for compositional data analyses will be mentioned in parantheses in the headers.</p>
<p>Check out these papers for more details on why compositional data analyses methods, and a comprehensive walkthrough:</p>
<ol type="a">
<li>Gloor, G. B., et al.&nbsp;(2017). “Microbiome Datasets Are Compositional: And This Is Not Optional.” Front Microbiol 8: 2224.</li>
<li>Quinn, T. P., et al.&nbsp;(2019). “A field guide for the compositional analysis of any-omics data.” Gigascience 8(9).</li>
</ol>
</section>
<section id="total-sum-scaling-tss" class="level4">
<h4 class="anchored" data-anchor-id="total-sum-scaling-tss">4.1. Total Sum-Scaling (TSS)</h4>
<p>With this method, the function transforms the feature table into relative abundance by dividing the number of total reads of each sample. For more info, see this <a href="https://rdrr.io/github/HuaZou/MyRtools/man/normalize-methods.html">link</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>asvtab.c.StdByTotM2 <span class="ot">&lt;-</span> <span class="fu">decostand</span>(asvtab.c.noX, <span class="at">method =</span> <span class="st">"total"</span>, <span class="at">MARGIN =</span> <span class="dv">2</span>) <span class="co"># MARGIN: 1 = rows, and 2 = columns of x.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We want the results in percentage of OTUs/ASVs per sample, so:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>asvtab.c.StdByTotM2.p <span class="ot">&lt;-</span> asvtab.c.StdByTotM2<span class="sc">*</span><span class="dv">100</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="presenceabsence" class="level4">
<h4 class="anchored" data-anchor-id="presenceabsence">4.2 Presence/Absence</h4>
<p>Self explanatory. Converts count of each ASV/OTU per sample to presence (1) or absence (0) across all samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">decostand</span>(asvtab.c, <span class="at">method =</span> <span class="st">"pa"</span>, <span class="at">MARGIN =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="hellinger-transformation" class="level4">
<h4 class="anchored" data-anchor-id="hellinger-transformation">4.3 Hellinger Transformation</h4>
<p>All values in a row are divided by the row sum. Then, the square root of the values are taken. Hellinger Transformations minimize effects of vastly different sample total abundances.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">decostand</span>(asvtab.c, <span class="at">method =</span> <span class="st">"hellinger"</span>, <span class="at">MARGIN =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="square-root-transformation" class="level4">
<h4 class="anchored" data-anchor-id="square-root-transformation">4.4 Square-root Transformation</h4>
<p>Also self explanatory. Takes the square root of each ASV/OTU and sample. Here we are going to take the square-root transformation of our total sum-scaled sample ASV table earlier. As explained above the purpose is to reduce the relative contribution of quantitatively dominant ASVs, which will influence the balance between contributions of dominant and less abundant species to the correlation matrix. Square root transformation is chosen here as it is a relatively mild form of transformation and there are more severe forms (e.g.&nbsp;Fourth root, Log X+1, P/A) which will more drastically increase the influence of less-abundant species.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>asvtab.c.StdByTotM2.p.SqRt <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(asvtab.c.StdByTotM2.p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="centred-log-ratio-clr-transformation-compositional" class="level4">
<h4 class="anchored" data-anchor-id="centred-log-ratio-clr-transformation-compositional">4.5 Centred Log-ratio (CLR) Transformation [Compositional]</h4>
<p>CLR is the most common method of transformation used in compositional data analyses introduced by Aitchison (1986). It uses the geometric mean of the sample vector as its reference.</p>
<p>Prior to doing a CLR transformation, zero count values, which are often prevalent especially in sparse data needs to be deleted, replaced or estimated. You can do that by considering the zero count values as point estimates. We’ll use the ‘cmultRepl’ function in from the zCompositions package to do this, but there are also other packages and methods to do this.</p>
<p>‘cmultRepl’ imputes zeros in compositional count data sets based on a Bayesian-multiplicative replacement (replace zeros with a small value)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>asvtab.c.no0 <span class="ot">&lt;-</span> <span class="fu">cmultRepl</span>(asvtab.c.noX, <span class="at">method =</span> <span class="st">"GBM"</span>, <span class="at">output =</span> <span class="st">"p-counts"</span>) <span class="co"># the ‘‘p-counts" option returns pseudo-counts instead of proportions</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can do the CLR transformation. We need sample as rows, asv as columns, and we convert it into a dataframe after that.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>clr.no0<span class="ot">&lt;-</span><span class="fu">as.data.frame</span>(<span class="fu">clr</span>(<span class="fu">t</span>(asvtab.c.no0)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Other log-ratio based transformation (that you can explore in your own time):</p>
<ul>
<li>additive log-ratio transformation (alr)</li>
<li>multi-additive log-ratio [malr] transformations:
<ol type="a">
<li>inter-quartile log-ratio (iqlr) transformation,</li>
<li>robust centered log-ratio (rclr)</li>
</ol></li>
<li>isometric log-ratio (ilr)</li>
</ul>
</section>
</section>
<section id="alpha-diversity" class="level1">
<h1>5. Alpha diversity</h1>
<p>Alpha diversity metrices will give us an indication of how diverse each sample is, such as richness (no. of species observed in each site) and evenness (distribution of abundance across the species in a community/site).</p>
<section id="sampling-effort" class="level4">
<h4 class="anchored" data-anchor-id="sampling-effort">Sampling effort</h4>
<p>First let’s plot some good old rarefaction curves. You want your sampling effort to be approaching or on the plateau. These curves will tend to the “true” sample richness value.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rarecurve</span>(<span class="fu">t</span>(asvtab.c), <span class="at">step=</span><span class="dv">50</span>, <span class="at">cex=</span><span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-49-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="alpha-diversity-1" class="level4">
<h4 class="anchored" data-anchor-id="alpha-diversity-1">Alpha diversity</h4>
<p>Now let’s calculate the alpha diversity of the samples using Hill’s Numbers. Because we are not yet reaching plateau for many samples, these results will be difficult to interpret.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>alpha_diversity <span class="ot">&lt;-</span> <span class="fu">data.frame</span> (</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">richness =</span> <span class="fu">hill_taxa</span>(asvtab.c.StdByTotM2, <span class="at">q =</span> <span class="dv">0</span>, <span class="at">MARGIN =</span> <span class="dv">2</span>, <span class="at">base =</span> <span class="fu">exp</span>(<span class="dv">1</span>)),</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">shannon  =</span> <span class="fu">hill_taxa</span>(asvtab.c.StdByTotM2, <span class="at">q =</span> <span class="dv">1</span>, <span class="at">MARGIN =</span> <span class="dv">2</span>, <span class="at">base =</span> <span class="fu">exp</span>(<span class="dv">1</span>)),</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Simpson_inv =</span> <span class="fu">hill_taxa</span>(asvtab.c.StdByTotM2, <span class="at">q =</span> <span class="dv">2</span>, <span class="at">MARGIN =</span> <span class="dv">2</span>, <span class="at">base =</span> <span class="fu">exp</span>(<span class="dv">1</span>))</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>alpha_diversity<span class="sc">$</span>Sample_ID <span class="ot">&lt;-</span> <span class="fu">row.names</span>(alpha_diversity)</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>alpha_diversity <span class="ot">&lt;-</span> <span class="fu">merge</span>(alpha_diversity, metad.N140.c , <span class="at">by=</span> <span class="st">"Sample_ID"</span> , <span class="at">all=</span>T)</span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a><span class="co"># We can calculate Simpson evenness by dividing the Simpson diversity by the richness</span></span>
<span id="cb68-11"><a href="#cb68-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-12"><a href="#cb68-12" aria-hidden="true" tabindex="-1"></a>alpha_diversity<span class="sc">$</span>eveness <span class="ot">&lt;-</span> alpha_diversity<span class="sc">$</span>Simpson_inv<span class="sc">/</span>alpha_diversity<span class="sc">$</span>richness</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Bar plot (bp)</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggbarplot</span>(alpha_diversity, <span class="at">x =</span> <span class="st">"Sample_ID"</span>, <span class="at">y =</span> <span class="st">"richness"</span>,</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">fill =</span> <span class="st">"Plastic"</span>,               <span class="co"># change fill color by cyl</span></span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">color =</span> <span class="st">"white"</span>,            <span class="co"># Set bar border colors to white</span></span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">palette =</span> <span class="st">"jco"</span>,            <span class="co"># jco journal color palett. see ?ggpar</span></span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">sort.val =</span> <span class="st">"asc"</span>,           <span class="co"># Sort the value in ascending order</span></span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">sort.by.groups =</span> <span class="cn">TRUE</span>,      <span class="co"># Sort inside each group</span></span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a>                <span class="at">x.text.angle =</span> <span class="dv">90</span>           <span class="co"># Rotate vertically x axis texts</span></span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggbarplot</span>(alpha_diversity, <span class="at">x =</span> <span class="st">"Sample_ID"</span>, <span class="at">y =</span> <span class="st">"eveness"</span>,</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>          <span class="at">fill =</span> <span class="st">"Plastic"</span>,               <span class="co"># change fill color by cyl</span></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>          <span class="at">color =</span> <span class="st">"white"</span>,            <span class="co"># Set bar border colors to white</span></span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">palette =</span> <span class="st">"jco"</span>,            <span class="co"># jco journal color palett. see ?ggpar</span></span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">sort.val =</span> <span class="st">"asc"</span>,           <span class="co"># Sort the value in ascending order</span></span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">sort.by.groups =</span> <span class="cn">TRUE</span>,      <span class="co"># Sort inside each group</span></span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">x.text.angle =</span> <span class="dv">90</span>           <span class="co"># Rotate vertically x axis texts</span></span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-51-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggbarplot</span>(alpha_diversity, <span class="at">x =</span> <span class="st">"Sample_ID"</span>, <span class="at">y =</span> <span class="st">"shannon"</span>,</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>          <span class="at">fill =</span> <span class="st">"Plastic"</span>,               <span class="co"># change fill color by cyl</span></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>          <span class="at">color =</span> <span class="st">"white"</span>,            <span class="co"># Set bar border colors to white</span></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">palette =</span> <span class="st">"jco"</span>,            <span class="co"># jco journal color palett. see ?ggpar</span></span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">sort.val =</span> <span class="st">"asc"</span>,           <span class="co"># Sort the value in ascending order</span></span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">sort.by.groups =</span> <span class="cn">TRUE</span>,      <span class="co"># Sort inside each group</span></span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">x.text.angle =</span> <span class="dv">90</span>           <span class="co"># Rotate vertically x axis texts</span></span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-51-3.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare Shannon diversity between plastic groups with a boxplot:</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggboxplot</span>(alpha_diversity, <span class="at">x =</span> <span class="st">"Plastic"</span>, <span class="at">y =</span> <span class="st">"shannon"</span>,</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">color =</span> <span class="st">"Plastic"</span>, <span class="at">palette =</span> <span class="st">"jco"</span>,</span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">add =</span> <span class="st">"jitter"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-51-4.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="phylogenetic-alpha-diversity" class="level4">
<h4 class="anchored" data-anchor-id="phylogenetic-alpha-diversity">Phylogenetic alpha Diversity</h4>
<p>If you have a tree for your OTUs/ASVs you can also calculate phylogeny diversity, i.e.&nbsp;if you have a community that is skewed towards a few branches of the tree or if it’s evenly distributed across the tree.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(picante)</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>df.pd <span class="ot">&lt;-</span> <span class="fu">pd</span>(<span class="fu">t</span>(otu_table_ps1), treefile_p1,<span class="at">include.root=</span>F) <span class="co"># t(ou_table) transposes the table for use in picante and the tre file comes from the first code chunck we used to read tree file (see making a phyloseq object section).</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="distance-metrices" class="level1">
<h1>6. Distance Metrices</h1>
<p>As shown in the flow chart above, before we ordinate our data in a chart, or perform multivariate analyses on it, we first need to define the resemblance (similarity/dissimilarity) between every pair of sample for all samples, which will result in a similarity/dissimilarity matrix.</p>
<p>There are many different resemblance indices, some of the common ones are: Bray-Curtis, Euclidean, Gower, Morista, Cao, Jaccard, etc…</p>
<p>For a comprehensive list of available indices:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>    ?<span class="fu">vegdist</span>() <span class="co"># in package vegan</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>    distanceMethodList <span class="co"># in package phyloseq</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We’ll use the Bray-Curtis index today, which is one of the most common:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>asvtab.c.StdByTotM2.p.SqRt.bray <span class="ot">&lt;-</span> <span class="fu">vegdist</span>(<span class="fu">t</span>(asvtab.c.StdByTotM2.p.SqRt), <span class="at">method =</span> <span class="st">"bray"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If your dataset only has a simple species list, with only presence or absence of each species, you could use the Jaccard matrix:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>asvtab.c.StdByTotM2.p.SqRt.jac <span class="ot">&lt;-</span> <span class="fu">vegdist</span>(<span class="fu">t</span>(asvtab.c.StdByTotM2.p), <span class="at">method =</span> <span class="st">"bray"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are also distance metrices that are more suitable for compositional analyses, such as the Aitchison distance matrix. We’ll compute resemblance matrix from centre-log ratio transformed ASV table earlier. The Aitchison distance based resemblance matrix can be computed using the package <i>coda.base</i>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>clr.no0.aitd <span class="ot">&lt;-</span> <span class="fu">dist</span>(clr.no0, <span class="at">method =</span> <span class="st">"aitchison"</span>) <span class="co"># aitchison distance will not work unless zeros in ASV table is imputed!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ordination-methods" class="level1">
<h1>7. Ordination Methods</h1>
<p>Ordination is representing data from a large number of sites as points in a multidimensional space.</p>
<p>Some of the ordination methods commonly used are:</p>
<p>Older-school:</p>
<ul>
<li><p>non-metric Multidimensional Scaling (nMDS)</p></li>
<li><p>Principle Coordinate Analyses (PCoA)</p></li>
<li><p>Correspondence Analyses (CA)</p></li>
<li><p>Principle Component Analyses (PCA)</p></li>
</ul>
<p>Newer-school:</p>
<ul>
<li><p>Uniform Manifold Approximation and Projection for Dimension Reduction (<a href="https://cran.r-project.org/web/packages/umap/vignettes/umap.html">UMAP</a>)</p></li>
<li><p>T-Distributed Stochastic Neighbor Embedding (<a href="https://cran.r-project.org/web/packages/tsne/index.html">t-SNE</a>)</p></li>
</ul>
<section id="nmds" class="level4">
<h4 class="anchored" data-anchor-id="nmds">7.1 nMDS</h4>
<p>nMDS constrains a distance matrix into a given number of dimensions, and aims to represent samples as points in a low dimensional space (usually 2 or 3). The closer the points are to each other the more similar thier community composition. You can use nMDS on a wide range of data types and resemblances. The nMDS we will build will be based on the Bray-Curtis resemblance matrix that we computed earlier:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>(nmds<span class="ot">&lt;-</span><span class="fu">metaMDS</span>(asvtab.c.StdByTotM2.p.SqRt.bray, <span class="at">k =</span> <span class="dv">2</span>)) <span class="co"># where k is the number of dimensions </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To plot the nMDS:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(nmds<span class="sc">$</span>points, <span class="at">type =</span> <span class="st">"n"</span>);<span class="fu">text</span>(nmds<span class="sc">$</span>points, <span class="at">label =</span> <span class="fu">rownames</span>(nmds<span class="sc">$</span>points))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-58-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Now that’s not a very pretty (or informative) plot, is it? Let’s make it better :) To do that, we first need to save the coordinates of our nMDS plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>nmds.coord <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(nmds<span class="sc">$</span>points)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Another useful value to take note of in an nMDS plot is its stress value. Stress values actually measure the values in the Shepard plot and gives an indication of how faithfully the high dimension relationships are represented in low dimensional ordinations. Usually we aim for a value of &lt; 0.1. The lower the better. Anything above 0.2-0.3 signifies unreliability. If the stress values are high, you may need to increase the number of dimensions during the nMDS analyses, since this signifies that your plot is too complex to be analyzed in just 2 dimensions. To view/save the stress value of our nMDS plot from earlier:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>nmds<span class="sc">$</span>stress</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1864234</code></pre>
</div>
</div>
<p>Now let’s make a nicer nMDS plot with the data we have saved, with the plastic type of each sample represented by different colors. To do so, we’ll first make our categorical variables as factors:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>plastic.type <span class="ot">&lt;-</span> <span class="fu">factor</span>(metad.N140.c.noX<span class="sc">$</span>Plastic[<span class="dv">1</span><span class="sc">:</span><span class="dv">38</span>], <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"PE"</span>, <span class="st">"PP"</span>, <span class="st">"PS"</span>), <span class="at">labels=</span><span class="fu">c</span>(<span class="st">'PE'</span>, <span class="st">'PP'</span>, <span class="st">'PS'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s make the plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>nmds.pretty <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(nmds.coord, <span class="fu">aes</span>(<span class="at">x =</span> MDS1, <span class="at">y =</span> MDS2)) <span class="sc">+</span> </span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">4</span>, <span class="fu">aes</span>(<span class="at">colour =</span> plastic.type))<span class="sc">+</span> </span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">size =</span> <span class="dv">12</span>, <span class="at">face =</span> <span class="st">"bold"</span>), </span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">face =</span> <span class="st">"bold"</span>, <span class="at">size =</span> <span class="dv">12</span>), </span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>, <span class="at">colour =</span><span class="st">"black"</span>), </span>
<span id="cb84-6"><a href="#cb84-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.12</span>, <span class="fl">0.85</span>), <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">face =</span> <span class="st">"bold"</span>, <span class="at">size =</span> <span class="dv">14</span>), </span>
<span id="cb84-7"><a href="#cb84-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">face =</span> <span class="st">"bold"</span>, <span class="at">size =</span> <span class="dv">14</span>, <span class="at">colour =</span> <span class="st">"black"</span>), </span>
<span id="cb84-8"><a href="#cb84-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">14</span>, <span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">face =</span> <span class="st">"bold"</span>), </span>
<span id="cb84-9"><a href="#cb84-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.background =</span> <span class="fu">element_blank</span>(), <span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">fill =</span> <span class="cn">NA</span>, <span class="at">size =</span> <span class="fl">1.2</span>),</span>
<span id="cb84-10"><a href="#cb84-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.key=</span><span class="fu">element_blank</span>()) <span class="sc">+</span> </span>
<span id="cb84-11"><a href="#cb84-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"NMDS1"</span>, <span class="at">colour =</span> <span class="st">"Plastic Type"</span>, <span class="at">y =</span> <span class="st">"NMDS2"</span>)  <span class="sc">+</span> </span>
<span id="cb84-12"><a href="#cb84-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"#0000FF"</span>, <span class="st">"#a62800"</span>, <span class="st">"#00ff00"</span>))  <span class="sc">+</span></span>
<span id="cb84-13"><a href="#cb84-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x=</span><span class="fl">1.7</span>, <span class="at">y=</span><span class="fl">1.5</span>, <span class="at">size =</span> <span class="dv">4</span>, <span class="at">label =</span> <span class="st">"2D stress: 0.1864"</span>) <span class="sc">+</span></span>
<span id="cb84-14"><a href="#cb84-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label_repel</span>(<span class="at">label=</span>metad.N140.c.noX<span class="sc">$</span>Sample, <span class="at">label.size=</span><span class="cn">NA</span>, <span class="at">box.padding =</span> <span class="fl">0.1</span>, </span>
<span id="cb84-15"><a href="#cb84-15" aria-hidden="true" tabindex="-1"></a>                   <span class="at">fill =</span> <span class="cn">NA</span>, <span class="at">min.segment.length =</span> <span class="cn">Inf</span>) <span class="co"># set min.segment.length to inf to remove line pointers</span></span>
<span id="cb84-16"><a href="#cb84-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-17"><a href="#cb84-17" aria-hidden="true" tabindex="-1"></a>nmds.pretty</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-62-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="pca-compositional" class="level4">
<h4 class="anchored" data-anchor-id="pca-compositional">7.2 PCA (Compositional)</h4>
<p>PCA is a projection of points in high dimensional variable space onto axes that minimizes residual variation in Euclidean space. It is typically used on datasets where there are not many zeros, so traditionally it was not considered very suitable for many biological data, though this has started to change in the recent years with the push for the compositional approach to data analyses. It is recommended to do your research on whether this is suitable for your dataset before proceeding!</p>
<p>Gloor et.al. (2017) recommends the use of Aitchison distance or the philr phylogenetic transform to define resemblance matrix. philr is based on binary partitions along an evolutionary tree and can be used as a UniFrac distance metric replacement, but only examines the relationships between chosen partitions. Aitchison distance is defined as the Euclidean distance between samples that are clr transformed.</p>
<p>Here we will compute our PCA based on the Aitchison distance, which we have already computed above.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>clr.no0.aitd.pca <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(clr.no0.aitd) <span class="co"># from centre-log ratio transformed ASV table earlier </span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(clr.no0.aitd.pca)</span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(clr.no0.aitd.pca)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can also visualize the eigenvalues/scree plot of the PCA. This shows the percentage of variances explained by each principle component.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>clr.no0.aitd.pca.eig <span class="ot">&lt;-</span> <span class="fu">fviz_eig</span>(clr.no0.aitd.pca) <span class="co"># need package factoextra</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>clr.no0.aitd.pca.eig</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To plot the PCA:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>pca.aitd <span class="ot">&lt;-</span> <span class="fu">fviz_pca_ind</span>(clr.no0.aitd.pca , </span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">repel =</span> <span class="cn">TRUE</span>,</span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">geom =</span> <span class="st">"point"</span>)</span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a>pca.aitd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="pcoa" class="level4">
<h4 class="anchored" data-anchor-id="pcoa">7.3 PCoA</h4>
<p>PCoA is similar to nMDS, except that it approximates eigenvectors based on the resemblance matrix. It is a projection of points onto axes that minimizes residual variation in the space of the resemblance measure. It is also very flexible and accepts any symmetrical resemblance matrix.</p>
<p>To learn more about the function in <i>vegan</i>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>?<span class="fu">wcmdscale</span>() <span class="co"># called weighted but here we do not weight our samples, so it's a normal PCoA</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>A simple script to plot PCoA:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>pcoa<span class="ot">&lt;-</span><span class="fu">wcmdscale</span>(asvtab.c.StdByTotM2.p.SqRt.bray, <span class="at">eig =</span> <span class="cn">TRUE</span>)</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pcoa)</span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a>pcoa.coords <span class="ot">&lt;-</span> pcoa<span class="sc">$</span>points <span class="co"># samples coordinates across eigenvectors</span></span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a>pcoa.coords <span class="ot">&lt;-</span> pcoa<span class="sc">$</span>eig <span class="co"># eigen values (quantity of information on each eigenvector)</span></span>
<span id="cb89-7"><a href="#cb89-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-8"><a href="#cb89-8" aria-hidden="true" tabindex="-1"></a>(pcoa<span class="sc">$</span>eig<span class="sc">/</span><span class="fu">sum</span>(pcoa<span class="sc">$</span>eig))<span class="sc">*</span><span class="dv">100</span> <span class="co"># % eigen values (quantity of information on each eigenvector)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="multivariate-comparisons" class="level1">
<h1>8. Multivariate Comparisons</h1>
<section id="permutational-multivariate-analysis-of-variance-using-distance-matrices-permanova" class="level4">
<h4 class="anchored" data-anchor-id="permutational-multivariate-analysis-of-variance-using-distance-matrices-permanova">8.2 Permutational Multivariate Analysis of Variance Using Distance Matrices (PERMANOVA)</h4>
<p>You have made your ordination plots, which show that there is some sort of grouping of your samples and variation between the groups. But how much variation is there really? Which pairs of groups (communities) are more similar to different to each other? Are they statistically significant?</p>
<p>In a typical Analysis of Variance (ANOVA) you study the variance of a variable across categories. PERMANOVA allows you to do this but for distance matrices across clusters/categories/variables or groups of samples. It’s non-parametrical and used more and more in microbial ecology. To learn more about the function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>?<span class="fu">adonis</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>An example of a one-way PERMANOVA with our dataset:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>permanova<span class="ot">&lt;-</span><span class="fu">adonis2</span>((<span class="fu">t</span>(asvtab.c.StdByTotM2.p.SqRt)) <span class="sc">~</span> Plastic , <span class="at">data =</span> metad.N140.c.noX, <span class="at">permutations =</span> <span class="dv">999</span>, <span class="at">method =</span> <span class="st">"bray"</span>)</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>permanova</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for adonis under reduced model
Terms added sequentially (first to last)
Permutation: free
Number of permutations: 999

adonis2(formula = (t(asvtab.c.StdByTotM2.p.SqRt)) ~ Plastic, data = metad.N140.c.noX, permutations = 999, method = "bray")
         Df SumOfSqs     R2      F Pr(&gt;F)    
Plastic   2   1.2293 0.1029 2.0072  0.001 ***
Residual 35  10.7173 0.8971                  
Total    37  11.9465 1.0000                  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>It may also be helpful to make pair-wise comparisons among all pairs of the factor of interest (i.e.&nbsp;the different type of Plastics). Unfortunately the stock ‘adonis2’ function in <i>vegan</i> does not do this, so we will have to rely on a separate package. Google will point you to many options, for today we will rely on the package <i>ecole</i></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"remotes"</span>)</span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"phytomosaic/ecole"</span>)</span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ecole)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>permanova_pw <span class="ot">&lt;-</span> <span class="fu">permanova_pairwise</span>(</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">t</span>(asvtab.c.StdByTotM2.p.SqRt),</span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">grp =</span> metad.N140.c.noX<span class="sc">$</span>Plastic,</span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">permutations =</span> <span class="dv">999</span>,</span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">method =</span> <span class="st">"bray"</span>,</span>
<span id="cb94-6"><a href="#cb94-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">padj =</span> <span class="st">"bonferroni"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>permanova_pw</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     pairs  SumOfSqs  F.Model         R2  pval p.adj
1 PE vs PP 0.6474769 2.087657 0.07432648 0.002 0.006
2 PE vs PS 0.7340274 2.538368 0.08894580 0.001 0.003
3 PP vs PS 0.4014957 1.234897 0.06420087 0.130 0.390</code></pre>
</div>
</div>
<p>From the PERMANOVA above, we can conclude that the bacterial communities are from the different plastic types are overall significantly different from each other. The pairwise PERMANOVA shows that plastic type PE contains a significantly different bacterial composition from PP and PS, but PP and PS does not differ significantly.</p>
<p>If you have more complex experimental designs or datasets, the interactions between the effects/factors can also be tested and teased apart with 2-way (or higher) crossed or nested PERMANOVA tests.</p>
</section>
<section id="test-of-homogeneity-of-dispersions-permdisp" class="level4">
<h4 class="anchored" data-anchor-id="test-of-homogeneity-of-dispersions-permdisp">8.3 Test of homogeneity of dispersions (PERMDISP)</h4>
<p>It is often easy to confound the differences between groups of samples and dispersion within the groups itself. Here, PERMDISP complements the PERMANOVA test by providing an indication of how much multivariate dispersion exists within a group of samples.</p>
<p>In <i>vegan</i>, we can do a PERMDISP with the ‘betadisper’ function. The test is typically done on Bray-Curtis resemblances between samples, though in theory the test can be used with any resemblance measure.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first define the factors:</span></span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>permd.grp <span class="ot">&lt;-</span> <span class="fu">factor</span>(metad.N140.c.noX<span class="sc">$</span>Plastic) </span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a>permdisp <span class="ot">&lt;-</span> <span class="fu">betadisper</span>(asvtab.c.StdByTotM2.p.SqRt.bray, permd.grp, </span>
<span id="cb97-5"><a href="#cb97-5" aria-hidden="true" tabindex="-1"></a>                       <span class="at">type =</span> <span class="st">"centroid"</span>, <span class="at">bias.adjust =</span> <span class="cn">FALSE</span>, <span class="at">sqrt.dist =</span> <span class="cn">FALSE</span>, <span class="at">add =</span> <span class="cn">FALSE</span>)</span>
<span id="cb97-6"><a href="#cb97-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb97-7"><a href="#cb97-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate variation p-value based on traditional tables</span></span>
<span id="cb97-8"><a href="#cb97-8" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(permdisp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: Distances
          Df   Sum Sq   Mean Sq F value Pr(&gt;F)
Groups     2 0.016442 0.0082209  1.9894  0.152
Residuals 35 0.144633 0.0041324               </code></pre>
</div>
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Permutation test for the F-values</span></span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a><span class="fu">permutest</span>(permdisp, <span class="at">pairwise =</span> <span class="cn">TRUE</span>, <span class="at">permutations =</span> <span class="dv">999</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Permutation test for homogeneity of multivariate dispersions
Permutation: free
Number of permutations: 999

Response: Distances
          Df   Sum Sq   Mean Sq      F N.Perm Pr(&gt;F)
Groups     2 0.016442 0.0082209 1.9894    999  0.159
Residuals 35 0.144633 0.0041324                     

Pairwise comparisons:
(Observed p-value below diagonal, permuted p-value above diagonal)
        PE      PP    PS
PE         0.08400 0.846
PP 0.08578         0.110
PS 0.83968 0.11186      </code></pre>
</div>
</div>
<p>Based on the PERMDISP analyses above, the H<sub>0</sub> hypothesis (of no difference in dispersion of the groups) is not rejected (p=0.152), so the significant PERMANOVA is caused by difference in plastic type (not by dispersion).</p>
</section>
<section id="analysis-of-similarity-anosim---non-parametric" class="level4">
<h4 class="anchored" data-anchor-id="analysis-of-similarity-anosim---non-parametric">8.4 Analysis of Similarity (ANOSIM - Non Parametric)</h4>
<p>ANOSIM also allows you to statistically test whether there are significant differences between groups of samples (e.g.&nbsp;different sites, treatment or sampling times). It is comparable to PERMANOVA, although you can only do up to two-way experimental designs. ANOSIM also ranks values in the resemblance matrix prior to the analyses. As compared to PERMANOVA, the ANOSIM R-statistic is scaled to range only between -1 and +1, which may be useful as an absolute measure of strength of differences between groups.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>anosim <span class="ot">&lt;-</span> <span class="fu">with</span>(metad.N140.c.noX, <span class="fu">anosim</span>(<span class="fu">t</span>(asvtab.c.StdByTotM2.p.SqRt.bray), Plastic))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(anosim)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="exploring-phylogenetic-diversity" class="level1">
<h1>9. Exploring Phylogenetic Diversity</h1>
<p>Finally, we would also like to know what is the distribution of the different phyla in the bacterial communities, and how this distribution may vary between the different plastic types.</p>
<section id="plotting-abundances-with-phyloseq" class="level4">
<h4 class="anchored" data-anchor-id="plotting-abundances-with-phyloseq">Plotting Abundances with <i>Phyloseq</i></h4>
<p><i>Phyloseq</i> is a great package to visualize your data. It makes sure that your data tables are nicely matching. So it’s fool proof! Here we cover some absolute basics. The <i>Phyloseq</i> graphs use the same aesthetics as ggplot.</p>
</section>
<section id="generate-your-phyloseq-objects-and-make-a-simple-plot" class="level4">
<h4 class="anchored" data-anchor-id="generate-your-phyloseq-objects-and-make-a-simple-plot">Generate your <i>Phyloseq</i> objects and make a simple plot</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>OTU <span class="ot">=</span> asvtab.c.StdByTotM2 <span class="sc">|&gt;</span> <span class="fu">as.matrix</span>() <span class="sc">|&gt;</span> <span class="fu">otu_table</span>(, <span class="at">taxa_are_rows =</span> <span class="cn">TRUE</span>)</span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>TAX <span class="ot">=</span> taxo.mat <span class="sc">|&gt;</span> <span class="fu">as.matrix</span>() <span class="sc">|&gt;</span> <span class="fu">tax_table</span>()</span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">=</span> metad.N140.c <span class="sc">|&gt;</span> <span class="st">`</span><span class="at">rownames&lt;-</span><span class="st">`</span>( metad.N140.c<span class="sc">$</span>Sample_ID ) <span class="sc">|&gt;</span> <span class="fu">sample_data</span>()</span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a>carbom <span class="ot">&lt;-</span> <span class="fu">phyloseq</span>(OTU, TAX, samples)</span>
<span id="cb103-6"><a href="#cb103-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-7"><a href="#cb103-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_bar</span>(carbom, <span class="at">fill =</span> <span class="st">"Phylum"</span>)<span class="sc">+</span></span>
<span id="cb103-8"><a href="#cb103-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">color=</span>Phylum, <span class="at">fill=</span>Phylum), <span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">position=</span><span class="st">"stack"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-76-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_bar</span>(carbom, <span class="at">fill=</span><span class="st">"Phylum"</span>) <span class="sc">+</span></span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">color=</span>Phylum, <span class="at">fill=</span>Phylum), <span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">position=</span><span class="st">"stack"</span>)<span class="sc">+</span></span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>Plastic, <span class="at">scales=</span><span class="st">"free_x"</span>, <span class="at">nrow=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-76-2.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="zooming-into-specific-taxa" class="level4">
<h4 class="anchored" data-anchor-id="zooming-into-specific-taxa">Zooming into specific taxa</h4>
<p>Let’s zoom into the Bacteriodetes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>GPr <span class="ot">=</span> <span class="fu">subset_taxa</span>(carbom, (Phylum <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Bacteroidota"</span>)))</span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_bar</span>(GPr, <span class="at">fill=</span><span class="st">"Family"</span>) <span class="sc">+</span></span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">color=</span>Family, <span class="at">fill=</span>Family), <span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">position=</span><span class="st">"stack"</span>)<span class="sc">+</span></span>
<span id="cb105-5"><a href="#cb105-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>Plastic, <span class="at">scales=</span><span class="st">"free_x"</span>, <span class="at">nrow=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-77-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Then we’ll plot the 100 most abundant Bacteroidetes ASVs:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>TopASVs <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">sort</span>(<span class="fu">taxa_sums</span>(GPr), <span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>])</span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a>Top <span class="ot">&lt;-</span> <span class="fu">prune_taxa</span>(TopASVs, GPr)</span>
<span id="cb106-4"><a href="#cb106-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb106-5"><a href="#cb106-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_bar</span>(Top, <span class="at">fill=</span><span class="st">"Genus"</span>) <span class="sc">+</span></span>
<span id="cb106-6"><a href="#cb106-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">color=</span>Genus, <span class="at">fill=</span>Genus), <span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">position=</span><span class="st">"stack"</span>)<span class="sc">+</span></span>
<span id="cb106-7"><a href="#cb106-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>Plastic, <span class="at">scales=</span><span class="st">"free_x"</span>, <span class="at">nrow=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-78-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="another-way-to-plot" class="level4">
<h4 class="anchored" data-anchor-id="another-way-to-plot">Another way to plot</h4>
<p><i>Phyloseq</i> plotting is good but will only take you so far. You can export the tables from <i>Phyloseq</i> into normal data.frames or generate ggplot ready tables for more flexible plotting.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Table by Phylum</span></span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a>y1 <span class="ot">&lt;-</span> <span class="fu">tax_glom</span>(carbom, <span class="at">taxrank =</span> <span class="st">'Phylum'</span>) <span class="co"># agglomerate taxa</span></span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a>y1 <span class="ot">&lt;-</span> <span class="fu">psmelt</span>(y1)</span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-5"><a href="#cb107-5" aria-hidden="true" tabindex="-1"></a><span class="co"># PLOT!</span></span>
<span id="cb107-6"><a href="#cb107-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-7"><a href="#cb107-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>y1, <span class="fu">aes</span>(<span class="at">x=</span>Sample, <span class="at">y=</span>Abundance, <span class="at">fill=</span>Phylum)) <span class="sc">+</span></span>
<span id="cb107-8"><a href="#cb107-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(), <span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">position=</span><span class="st">"stack"</span>)<span class="sc">+</span></span>
<span id="cb107-9"><a href="#cb107-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>Plastic, <span class="at">scales=</span><span class="st">"free_x"</span>, <span class="at">nrow=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="index_files/figure-html/unnamed-chunk-79-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="more-interesting-materials-and-references" class="level1">
<h1>More interesting materials and references</h1>
<ol type="1">
<li><p>Learn R multivariate data analysis from one of Vegan’s developers: <a href="https://www.youtube.com/live/tVnnG7mFeqA?feature=share">Video1</a> | <a href="https://www.youtube.com/live/PR1B_JkO49s?feature=share">Video 2</a></p></li>
<li><p><a href="https://www.youtube.com/watch?v=D6CunpqF04E&amp;list=PLmNrK_nkqBpIIRdQTS2aOs5OD7vVMKWAi&amp;ab_channel=RiffomonasProject">Riffomonas Project microbiome analysis in R</a></p></li>
</ol>
<section id="session-info" class="level4">
<h4 class="anchored" data-anchor-id="session-info">Session Info</h4>
<details>
<summary>
Show full session info
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.2.2 (2022-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Monterey 12.5.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRlapack.dylib

locale:
[1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] corrplot_0.92         factoextra_1.0.7      ggpubr_0.6.0         
 [4] ggrepel_0.9.3         rstatix_0.7.2         hillR_0.5.1          
 [7] ALDEx2_1.30.0         ecole_0.9-2021        zCompositions_1.4.0-1
[10] truncnorm_1.0-8       NADA_1.6-1.1          survival_3.5-5       
[13] MASS_7.3-58.3         Hmisc_5.0-1           pgirmess_2.0.0       
[16] labdsv_2.0-1          mgcv_1.8-42           nlme_3.1-162         
[19] FactoMineR_2.7        ade4_1.7-22           compositions_2.0-5   
[22] coda.base_0.5.4.3     phyloseq_1.42.0       vegan_2.6-4          
[25] lattice_0.20-45       permute_0.9-7         lubridate_1.9.2      
[28] forcats_1.0.0         stringr_1.5.0         dplyr_1.1.0          
[31] purrr_1.0.1           readr_2.1.4           tidyr_1.3.0          
[34] tibble_3.2.0          ggplot2_3.4.1         tidyverse_2.0.0      
[37] biomformat_1.26.0    

loaded via a namespace (and not attached):
  [1] backports_1.4.1             plyr_1.8.8                 
  [3] igraph_1.4.1                sp_1.6-0                   
  [5] splines_4.2.2               BiocParallel_1.32.5        
  [7] GenomeInfoDb_1.34.9         digest_0.6.31              
  [9] foreach_1.5.2               htmltools_0.5.4            
 [11] fansi_1.0.4                 magrittr_2.0.3             
 [13] checkmate_2.1.0             cluster_2.1.4              
 [15] tzdb_0.3.0                  Biostrings_2.66.0          
 [17] matrixStats_0.63.0          bayesm_3.1-5               
 [19] timechange_0.2.0            colorspace_2.1-0           
 [21] xfun_0.37                   crayon_1.5.2               
 [23] RCurl_1.98-1.10             jsonlite_1.8.4             
 [25] iterators_1.0.14            ape_5.7-1                  
 [27] glue_1.6.2                  gtable_0.3.1               
 [29] zlibbioc_1.44.0             emmeans_1.8.5              
 [31] XVector_0.38.0              DelayedArray_0.24.0        
 [33] car_3.1-1                   RcppZiggurat_0.1.6         
 [35] Rhdf5lib_1.20.0             BiocGenerics_0.44.0        
 [37] DEoptimR_1.0-11             abind_1.4-5                
 [39] scales_1.2.1                mvtnorm_1.1-3              
 [41] DBI_1.1.3                   Rcpp_1.0.10                
 [43] xtable_1.8-4                spData_2.2.2               
 [45] htmlTable_2.4.1             units_0.8-1                
 [47] flashClust_1.01-2           foreign_0.8-84             
 [49] spdep_1.2-8                 proxy_0.4-27               
 [51] Formula_1.2-5               stats4_4.2.2               
 [53] DT_0.27                     htmlwidgets_1.6.1          
 [55] wk_0.7.1                    ellipsis_0.3.2             
 [57] farver_2.1.1                pkgconfig_2.0.3            
 [59] nnet_7.3-18                 multcompView_0.1-8         
 [61] deldir_1.0-6                utf8_1.2.3                 
 [63] labeling_0.4.2              tidyselect_1.2.0           
 [65] rlang_1.0.6                 reshape2_1.4.4             
 [67] munsell_0.5.0               tools_4.2.2                
 [69] cli_3.6.0                   generics_0.1.3             
 [71] broom_1.0.4                 evaluate_0.20              
 [73] fastmap_1.1.1               yaml_2.3.7                 
 [75] knitr_1.42                  robustbase_0.95-0          
 [77] s2_1.1.2                    leaps_3.1                  
 [79] compiler_4.2.2              rstudioapi_0.14            
 [81] ggsignif_0.6.4              e1071_1.7-13               
 [83] stringi_1.7.12              Matrix_1.5-3               
 [85] classInt_0.4-9              ggsci_3.0.0                
 [87] tensorA_0.36.2              multtest_2.54.0            
 [89] vctrs_0.5.2                 pillar_1.8.1               
 [91] lifecycle_1.0.3             rhdf5filters_1.10.0        
 [93] estimability_1.4.1          data.table_1.14.8          
 [95] bitops_1.0-7                GenomicRanges_1.50.2       
 [97] R6_2.5.1                    KernSmooth_2.23-20         
 [99] gridExtra_2.3               IRanges_2.32.0             
[101] codetools_0.2-19            boot_1.3-28.1              
[103] SummarizedExperiment_1.28.0 rhdf5_2.42.0               
[105] withr_2.5.0                 S4Vectors_0.36.2           
[107] GenomeInfoDbData_1.2.9      parallel_4.2.2             
[109] hms_1.1.2                   grid_4.2.2                 
[111] rpart_4.1.19                coda_0.19-4                
[113] class_7.3-21                Rfast_2.0.7                
[115] rmarkdown_2.20              carData_3.0-5              
[117] MatrixGenerics_1.10.0       Rtsne_0.16                 
[119] sf_1.0-10                   scatterplot3d_0.3-42       
[121] Biobase_2.58.0              base64enc_0.1-3            </code></pre>
</div>
</div>
</details>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>